File: Android/shared/src/main/kotlin/shared/dto/objects/procedure/Service.kt
@file:UseSerializers(
    UUIDSerializer::class,
)

package shared.dto.objects.procedure

import kotlinx.serialization.Serializable
import kotlinx.serialization.UseSerializers
import shared.dto.enums.ServiceTags
import shared.dto.primitives.Pagination
import shared.dto.primitives.TranslatedServiceTag
import shared.dto.protocols.Parametable
import shared.dto.protocols.Responsable
import shared.serializers.UUIDSerializer
import java.util.UUID

object Service {
    data object Parameters {
        @Serializable
        data class All(
            val salonsFilter: List<UUID>? = null,
            val employeesFilter: List<UUID>? = null,
            val valueFilter: String? = null,
            val pagination: Pagination? = null,
        ) : Parametable()

        @Serializable
        data class Create(
            val title: String,
            val tags: Array<ServiceTags>,
        ) : Parametable()

        @Serializable
        data class Update(
            val title: String? = null,
            val tags: Array<ServiceTags>? = null,
        ) : Parametable()
    }

    data object Responses {
        @Serializable
        data class Create(
            val id: UUID,
            val title: String,
            val tags: Array<TranslatedServiceTag>,
        ) : Responsable

        @Serializable
        data class Update(
            val id: UUID,
            val title: String,
            val tags: Array<TranslatedServiceTag>,
        ) : Responsable

        @Serializable
        data class All(
            val services: List<Helpers.Service>,
        ) : Responsable

        @Serializable
        data class Retrieve(
            val id: UUID,
            val title: String,
            val tags: Array<TranslatedServiceTag>,
        ) : Responsable

        data object Helpers {

            @Serializable
            data class Service(
                val id: UUID,
                val title: String,
                val tags: Array<TranslatedServiceTag>,
            ) : Responsable
        }
    }
}

File: IOS/DTOs/Objects/MaestriProcedure/Service.swift
import Foundation
import MemberwiseInit

public enum Service {
    public enum Helpers {}
    
    public enum Parameters {}
    public enum Responses {}
}

// MARK: - Parameters -

public extension Service.Parameters {
    
    @MemberwiseInit(.public)
    struct All: Parametable {
        public let salonsFilter: [UUID]?
        public let employeesFilter: [UUID]?
        public let valueFilter: String?
        public let pagination: Pagination?
    }

    @MemberwiseInit(.public)
    struct Create: Parametable {
        public let title: String
        public let tags: [ServiceTags]
    }

    @MemberwiseInit(.public)
    struct Update: Parametable {
        public let title: String?
        public let tags: [ServiceTags]?
    }
}

// MARK: - Responses -

public extension Service.Responses {
    
    @MemberwiseInit(.public)
    struct Create: Responsable {
        public let id: UUID
        public let title: String
        public let tags: [TranslatedServiceTag]
    }
    
    @MemberwiseInit(.public)
    struct Update: Responsable {
        public let id: UUID
        public let title: String
        public let tags: [TranslatedServiceTag]
    }
    
    @MemberwiseInit(.public)
    struct All: Responsable {
        public let services: [Service.Helpers.ServiceResponse]
    }
    
    @MemberwiseInit(.public)
    struct Retrieve: Responsable {
        public let id: UUID
        public let title: String
        public let parameters: [Service.Helpers.Parameter]
        public let tags: [TranslatedServiceTag]
    }
}

public extension Service.Helpers {
    
    @MemberwiseInit(.public)
    struct ServiceResponse: Codable {
        public let id: UUID
        public let title: String
        public let tags: [TranslatedServiceTag]    }
    
    @MemberwiseInit(.public)
    struct Parameter: Codable {
        public let id: UUID
        public let title: String
        public let cases: [Case]
    }

    @MemberwiseInit(.public)
    struct Case: Codable {
        public let id: Int
        public let title: String
    }
}
